cheddar/src/internal.rs

```diff
@@ -139,27 +130,36 @@ impl Contract {
 
         if unused_amount > 0 {
             let receiver_balance = self.accounts.get(&receiver_id).unwrap_or(0);
+            // receiver has positive balance, so we can refund.
             if receiver_balance > 0 {
-                let refund_amount = std::cmp::min(receiver_balance, unused_amount);
+                // adjust the refund amount to the receiver balance
+                unused_amount = std::cmp::min(receiver_balance, unused_amount);
                 self.accounts
-                    .insert(&receiver_id, &(receiver_balance - refund_amount));
+                    .insert(&receiver_id, &(receiver_balance - unused_amount));
 
-                if let Some(sender_balance) = self.accounts.get(&sender_id) {
+                // now we will try to update sender balance
+                if let Some(sender_balance) = self.accounts.get(sender_id) {
                     self.accounts
-                        .insert(&sender_id, &(sender_balance + refund_amount));
+                        .insert(sender_id, &(sender_balance + unused_amount));
                     log!(
                         "Refund {} from {} to {}",
-                        refund_amount,
+                        unused_amount,
                         receiver_id,
                         sender_id
                     );
-                    return (amount - refund_amount, 0);
+                    return (amount - unused_amount, 0);
                 } else {
                     // Sender's account was deleted, so we need to burn tokens.
-                    self.total_supply -= refund_amount;
-                    log!("The account of the sender was deleted");
-                    return (amount, refund_amount);
```

Comment:
the solution was to put `amount - refund_amount` here.
Not change the entire fn

Reply:
the change in this function is:
+ reuse `unused_amount` -- as we fix it so no need to confuse the reader with a new variable
+ fix the return value
+ added logs to explain to the user that we were not able to correctly make a refund.

Reply:
no problem, let's do it this way, but it's easier for the reader if you use immutable variables, that's the reason why that's the default. It's way easier to read and reason about the code if vars are immutable.